frontend/apps/web/components/RunTImeline/RunTimeline.tsx

```diff
@@ -0,0 +1,474 @@
+import { JobRunEvent, JobRunStatus } from '@neosync/sdk';
+
+import { JobRunStatus as JobRunStatusEnum } from '@neosync/sdk';
+import {
+  CheckCircledIcon,
+  CrossCircledIcon,
+  MinusCircledIcon,
+  MixerHorizontalIcon,
+} from '@radix-ui/react-icons';
+import {
+  addMilliseconds,
+  format,
+  formatDuration,
+  intervalToDuration,
+} from 'date-fns';
+import { ReactElement, useMemo, useState } from 'react';
+import Spinner from '../Spinner';
+import TruncatedText from '../TruncatedText';
+import { Badge } from '../ui/badge';
+import { Button } from '../ui/button';
+
+interface Props {
+  tasks: JobRunEvent[];
+  jobStatus?: JobRunStatusEnum;
+}
+
+type RunStatus = 'running' | 'completed' | 'failed' | 'canceled';
+
+export default function RunTimeline(props: Props): ReactElement {
+  const { tasks, jobStatus } = props;
+
+  const [selectedStatuses, setSelectedStatuses] = useState<RunStatus[]>([
+    'running',
+    'completed',
+    'failed',
+    'canceled',
+  ]);
+
+  const { timelineStart, totalDuration, timeLabels } = useMemo(() => {
+    // find earliest start date out of all of the activities
+    let startTime = Infinity;
+    let endTime = -Infinity;
+
+    tasks.forEach((t) => {
+      startTime = Math.min(
+        startTime,
+        convertTimestampToDate(t.startTime).getTime()
+      );
+
+      const errorDate = getCloseOrErrorOrCancelDate(t);
+      endTime = Math.max(
+        endTime,
+        errorDate.getTime(),
+        convertTimestampToDate(t.closeTime || t.startTime).getTime()
+      );
+    });
+
+    const start = new Date(startTime);
+    const end = new Date(endTime);
+
+    let duration = end.getTime() - start.getTime();
+
+    // add padding, but limit it to a maximum of 100ms on each side so we can see the entire timeline in view in the graph
+    const padding = Math.min(duration * 0.1, 300);
+    const adjustedStart = new Date(start.getTime() - padding);
+    const adjustedEnd = new Date(end.getTime() + padding);
+    const adjustedDuration = adjustedEnd.getTime() - adjustedStart.getTime();
+
+    const labelCount = 5;
+
+    const labels: Date[] = Array.from({ length: labelCount }, (_, i) =>
+      addMilliseconds(adjustedStart, (adjustedDuration * i) / (labelCount - 1))
+    );
+
+    return {
+      timelineStart: adjustedStart,
+      timelineEnd: adjustedEnd,
+      totalDuration: adjustedDuration,
+      timeLabels: labels,
+    };
+  }, [tasks]);
+
+  // handles filtering the tasks when the tasks or filters change
+  const filteredTasks = useMemo(() => {
+    return tasks.filter((task) => {
+      const status = getTaskStatus(task, jobStatus);
+      return selectedStatuses.includes(status);
+    });
+  }, [tasks, selectedStatuses, jobStatus]);
+
+  const handleStatusFilterChange = (status: RunStatus, checked: boolean) => {
+    setSelectedStatuses((prev) =>
+      checked ? [...prev, status] : prev.filter((s) => s !== status)
+    );
+  };
+
+  return (
```

Comment:
Oh, the last thing that could be nice is to just overall factor out this component to be a little more readable.

like it would be great if we had more named components here that more closely reflect the layout of the page.

Could definitely be more of a cleanup task as you can go pretty deep into that. But just wanted to comment on it as it can also help with separation of concerns and having discrete components that do specific things.

Def not a blocker.